components/resource_control/src/resource_group.rs

```diff
@@ -0,0 +1,400 @@
+        if DEFAULT_RESOURCE_GROUP_NAME == group_name {
+            return;
+        }
+        self.registry.lock().unwrap().iter().for_each(|controller| {
+            controller.remove_resource_group(group_name.as_bytes());
+        });
+        self.resource_groups.remove(&group_name);
+    }
+
+    pub fn get_resource_group(&self, name: &str) -> Option<Ref<'_, String, ResourceGroup>> {
+        self.resource_groups.get(&name.to_ascii_lowercase())
+    }
+
+    pub fn get_all_resource_groups(&self) -> Vec<ResourceGroup> {
+        self.resource_groups.iter().map(|g| g.clone()).collect()
+    }
+
+    pub fn derive_controller(&self, name: String, is_read: bool) -> Arc<ResourceController> {
+        let controller = Arc::new(ResourceController::new(name, is_read));
+        self.registry.lock().unwrap().push(controller.clone());
+        for g in &self.resource_groups {
+            let priority_factor = self.gen_group_priority_factor(g.value(), is_read);
+            controller.add_resource_group(g.key().clone().into_bytes(), priority_factor);
+        }
+
+        controller
+    }
+
+    pub fn advance_min_virtual_time(&self) {
+        for controller in self.registry.lock().unwrap().iter() {
+            controller.update_min_virtual_time();
+        }
+    }
+}
+
+pub struct ResourceController {
+    // resource controller name is not used currently.
+    #[allow(dead_code)]
+    name: String,
+    // We handle the priority differently between read and write request:
+    // 1. the priority factor is calcualte based on read/write RU settings.
+    // 2. for read request, we increase a constant virtual time delta at each `get_priority` call
+    //    because the cost can't be calculated at start, so we only increase a constant delta and
+    //    increase the real cost after task is executed; but don't increase it at write because
+    //    the cost is known so we just pre-consume it.
+    is_read: bool,
+    // record consumption of each resource group, name --> resource_group
+    resource_consumptions: DashMap<Vec<u8>, GroupPriorityTracker>,
+    last_min_vt: AtomicU64,
+}
+
+impl ResourceController {
+    pub fn new(name: String, is_read: bool) -> Self {
+        let controller = Self {
+            name,
+            is_read,
+            resource_consumptions: DashMap::new(),
+            last_min_vt: AtomicU64::new(0),
+        };
+        // add the "default" resource group
+        controller.add_resource_group(DEFAULT_RESOURCE_GROUP_NAME.as_bytes().to_owned(), 1);
+        controller
+    }
+
+    fn add_resource_group(&self, name: Vec<u8>, priority_factor: u64) {
+        let vt_delta_for_get = if self.is_read {
+            DEFAULT_PRIORITY_PER_READ_TASK * priority_factor
+        } else {
+            0
+        };
+        let group = GroupPriorityTracker {
+            weight: priority_factor,
+            virtual_time: AtomicU64::new(self.last_min_vt.load(Ordering::Acquire)),
+            vt_delta_for_get,
```

Comment:
Is write will use field? otherwise no need `is_read` field, so will two resource control for one resource group?

Reply:
Write and read uses different ResouceController(and GroupPriorityTracker), only their config are the same.